test/ssl/client_ssl_authentication_test.go

```diff
@@ -0,0 +1,226 @@
+	client2Key    = "client2-key.pem"
+	serverName    = "foo.bar.com"
+	server1CA     = "server1.pem"
+	server2CA     = "server2.pem"
+	clientCertPw  = "cert.pem"
+	clientKeyPw   = "privkey.pem"
+	password      = "foobar"
+)
+
+func TestMain(m *testing.M) {
+	rc, err := rc.NewRemoteControllerClient("localhost:9701")
+	remoteController = rc
+	if remoteController == nil || err != nil {
+		log.Fatal("create remote controller failed:", err)
+	}
+	m.Run()
+}
+
+func createMemberWithXML(path string) (clusterID string, err error) {
+	config, err := test.Read(path)
+	if err != nil {
+		return "", err
+	}
+	cluster, err := remoteController.CreateCluster("", config)
+	if err != nil {
+		return "", err
+	}
+	remoteController.StartMember(cluster.ID)
+	return cluster.ID, nil
+}
+
+func createClientConfigWithSSLConfig(clientCertPath string, clientKeyPath string, caPath string) (*config.Config, error) {
+	config := hazelcast.NewConfig()
+	sslConfig := config.NetworkConfig().SSLConfig()
+	sslConfig.SetEnabled(true)
+	err := sslConfig.SetCaPath(caPath)
+	if err != nil {
+		return nil, err
+	}
+	err = sslConfig.AddClientCertAndKeyPath(clientCertPath, clientKeyPath)
+	if err != nil {
+		return nil, err
+	}
+	sslConfig.ServerName = serverName
+	return config, nil
+}
+
+func TestSSLMutualAuthenticationConnect(t *testing.T) {
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client1Cert,
+		client1Key, server1CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client, err := hazelcast.NewClientWithConfig(config)
+	defer client.Shutdown()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestSSLMutualAuthentication_ClientDoesntKnowServerFail(t *testing.T) {
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client1Cert,
+		client1Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
+
+func TestSSLMutualAuthentication_ServerDoesntKnowClientFail(t *testing.T) {
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client2Cert,
+		client2Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
+
+func TestSSLMutualAuthentication_NeitherServerNorClientKnowsTheOtherFail(t *testing.T) {
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig("client2-cert.pem",
```

Comment:
maybe you can change client2-cert.pem with client2Cert for consistency across the tests